File: scripts/lib/audit_logger.py
#!/usr/bin/env python3
# Copyright (C) 2026 Moko Consulting <hello@mokoconsulting.tech>
#
# This file is part of a Moko Consulting project.
#
# SPDX-License-Identifier: GPL-3.0-or-later
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# (./LICENSE).
#
# FILE INFORMATION
# DEFGROUP: MokoStandards.Scripts
# INGROUP: MokoStandards.Library
# REPO: https://github.com/mokoconsulting-tech/MokoStandards
# FILE: scripts/lib/audit_logger.py
# VERSION: 03.01.03
# BRIEF: Enterprise audit trail system for MokoStandards scripts
# PATH: /scripts/lib/audit_logger.py
# NOTE: Provides structured logging with transaction tracking and SIEM integration

"""
Audit Logger for MokoStandards Scripts

Provides enterprise-grade audit trail with:
- Structured JSON logging
- Transaction ID tracking
- Security event logging
- SIEM integration via syslog
- Compliance reporting
"""

import json
import os
import sys
import syslog
import getpass
from datetime import datetime, date
from pathlib import Path
from typing import Optional, Dict, Any, List
from dataclasses import dataclass, field, asdict
from uuid import uuid4


# ============================================================
# Audit Event Data Structure
# ============================================================

@dataclass
class AuditEvent:
    """Audit event data structure"""
    event_id: str
    timestamp: str
    session_id: str
    component: str
    operation: str
    target: str
    user: str
    status: str
    metadata: Dict[str, Any] = field(default_factory=dict)

    def to_json(self) -> str:
        """Convert to JSON string"""
        return json.dumps(asdict(self))

    def to_syslog_message(self) -> str:
        """Convert to syslog message format"""
        return (f"[MokoStandards] {self.component}: {self.operation} "
                f"on {self.target} by {self.user} - {self.status}")


# ============================================================
# Audit Logger
# ============================================================

class AuditLogger:
    """Enterprise audit logging system"""

    def __init__(self, component: str, audit_dir: Optional[Path] = None,
                 enable_syslog: bool = True):
        """
        Initialize audit logger

        Args:
            component: Component name (script name)
            audit_dir: Directory for audit logs (default: ~/.mokostandards/logs)
            enable_syslog: Enable syslog integration
        """
        self.component = component
        self.enable_syslog = enable_syslog
        self.session_id = self._generate_session_id()
        self.events: List[AuditEvent] = []

        # Set up audit directory
        if audit_dir is None:
            audit_dir = Path.home() / ".mokostandards" / "logs"
        self.audit_dir = Path(audit_dir).expanduser()
        self.audit_dir.mkdir(parents=True, exist_ok=True)

        # Initialize syslog if enabled
        if self.enable_syslog:
            try:
                syslog.openlog(
                    ident="mokostandards",
                    facility=syslog.LOG_LOCAL0
                )
            except Exception as e:
                print(f"⚠️  Warning: Could not initialize syslog: {e}",
                      file=sys.stderr)
                self.enable_syslog = False

        # Log session start
        self.log_operation(
            operation="session_start",
            target=component,
            status="started",
            metadata={"session_id": self.session_id}
        )

    def _generate_session_id(self) -> str:
        """Generate unique session ID"""
        return f"{datetime.utcnow().strftime('%Y%m%d_%H%M%S')}_{uuid4().hex[:8]}"

    def _get_current_user(self) -> str:
        """Get current username"""
        try:
            return getpass.getuser()
        except Exception:
            return os.environ.get('USER', 'unknown')

    def log_operation(
        self,
        operation: str,
        target: str,
        user: Optional[str] = None,
        status: str = "started",
        metadata: Optional[Dict[str, Any]] = None
    ) -> str:
        """
        Log security-relevant operation

        Args:
            operation: Operation name (e.g., "bulk_update", "validate")
            target: Target of operation (e.g., "org:mokoconsulting-tech", "file.py")
            user: User performing operation (auto-detected if None)
            status: Operation status (started, success, failed)
            metadata: Additional context data

        Returns:
            Event ID for tracking
        """
        event = AuditEvent(
            event_id=uuid4().hex,
            timestamp=datetime.utcnow().isoformat() + "Z",
            session_id=self.session_id,
            component=self.component,
            operation=operation,
            target=target,
            user=user or self._get_current_user(),
            status=status,
            metadata=metadata or {}
        )

        self.events.append(event)
        self._write_to_json(event)

        if self.enable_syslog:
            self._write_to_syslog(event)

        return event.event_id

    def _write_to_json(self, event: AuditEvent):
        """Append event to JSON audit log"""
        try:
            log_file = self.audit_dir / f"{date.today()}.audit.json"
            with log_file.open("a") as f:
                f.write(event.to_json() + "\n")
        except Exception as e:
            print(f"⚠️  Warning: Failed to write audit log: {e}",
                  file=sys.stderr)

    def _write_to_syslog(self, event: AuditEvent):
        """Send event to syslog for SIEM integration"""
        try:
            # Map status to syslog priority
            priority_map = {
                "started": syslog.LOG_INFO,
                "success": syslog.LOG_INFO,
                "failed": syslog.LOG_ERR,
                "error": syslog.LOG_ERR,
                "warning": syslog.LOG_WARNING
            }
            priority = priority_map.get(event.status, syslog.LOG_INFO)

            syslog.syslog(priority, event.to_syslog_message())
        except Exception as e:
            print(f"⚠️  Warning: Failed to write to syslog: {e}",
                  file=sys.stderr)

    def log_success(self, operation: str, target: str,
                   metadata: Optional[Dict[str, Any]] = None) -> str:
        """Log successful operation"""
        return self.log_operation(operation, target, status="success", metadata=metadata)

    def log_failure(self, operation: str, target: str, error: str,
                   metadata: Optional[Dict[str, Any]] = None) -> str:
        """Log failed operation"""
        meta = metadata or {}
        meta['error'] = str(error)
        return self.log_operation(operation, target, status="failed", metadata=meta)

    def get_session_summary(self) -> Dict[str, Any]:
        """Get summary of session events"""
        status_counts = {}
        for event in self.events:
            status_counts[event.status] = status_counts.get(event.status, 0) + 1

        return {
            "session_id": self.session_id,
            "component": self.component,
            "event_count": len(self.events),
            "status_summary": status_counts,
            "start_time": self.events[0].timestamp if self.events else None,
            "end_time": self.events[-1].timestamp if self.events else None
        }

    def close(self):
        """Close audit session"""
        self.log_operation(
            operation="session_end",
            target=self.component,
            status="completed",
            metadata=self.get_session_summary()
        )

        if self.enable_syslog:
            try:
                syslog.closelog()
            except Exception as e:
                # Log error closing syslog but don't raise as we're already closing
                import sys
                print(f"Warning: Error closing syslog: {e}", file=sys.stderr)

    def __enter__(self):
        """Context manager entry"""
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        """Context manager exit"""
        if exc_type is not None:
            self.log_failure(
                operation="session_exception",
                target=self.component,
                error=f"{exc_type.__name__}: {exc_val}"
            )
        self.close()


# ============================================================
# Convenience Functions
# ============================================================

def create_audit_logger(component: str, audit_dir: Optional[Path] = None) -> AuditLogger:
    """Create audit logger instance"""
    return AuditLogger(component, audit_dir)


# ============================================================
# Example Usage
# ============================================================

if __name__ == "__main__":
    import argparse

    parser = argparse.ArgumentParser(description="Test MokoStandards Audit Logger")
    parser.add_argument('--component', default='test_script',
                        help='Component name')
    parser.add_argument('--test', action='store_true',
                        help='Run test audit operations')
    parser.add_argument('--show-logs', action='store_true',
                        help='Show recent audit logs')

    args = parser.parse_args()

    if args.test:
        print("🔍 Testing audit logger...")

        with create_audit_logger(args.component) as audit:
            # Test various operations
            audit.log_operation(
                operation="test_operation",
                target="test_target",
                status="started",
                metadata={"test_data": "value"}
            )

            audit.log_success(
                operation="test_success",
                target="success_target",
                metadata={"result": "ok"}
            )

            audit.log_failure(
                operation="test_failure",
                target="failure_target",
                error="Simulated error",
                metadata={"error_code": 500}
            )

            summary = audit.get_session_summary()
            print(f"✅ Test completed. Session ID: {summary['session_id']}")
            print(f"📊 Events logged: {summary['event_count']}")
            print(f"📁 Logs saved to: {audit.audit_dir}")

    elif args.show_logs:
        log_dir = Path.home() / ".mokostandards" / "logs"
        if log_dir.exists():
            print(f"📁 Audit logs in: {log_dir}")
            for log_file in sorted(log_dir.glob("*.audit.json")):
                print(f"  - {log_file.name}")
        else:
            print("ℹ️  No audit logs found")

    else:
        parser.print_help()

File: src/Enterprise/AuditLogger.php
<?php

declare(strict_types=1);

/**
 * Enterprise Audit Library - Structured audit logging for all operations.
 *
 * This class provides enterprise-grade audit logging capabilities with:
 * - Structured JSON logging to audit database
 * - Transaction ID tracking across operations
 * - Security event logging (who, what, when, where)
 * - Audit log rotation and archival
 * - Compliance reporting capabilities
 *
 * Copyright (C) 2026 Moko Consulting <hello@mokoconsulting.tech>
 *
 * SPDX-License-Identifier: GPL-3.0-or-later
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * @package MokoStandards\Enterprise
 * @version 03.02.00
 * @author MokoStandards Team
 * @license GPL-3.0-or-later
 */

namespace MokoStandards\Enterprise;

use DateTime;
use DateTimeZone;
use RuntimeException;

/**
 * Enterprise audit logger with transaction tracking and structured logging.
 *
 * Features:
 * - Transaction ID tracking
 * - Security event logging
 * - Structured JSON output
 * - Automatic log rotation
 * - Context manager support
 *
 * Example:
 * ```php
 * $logger = new AuditLogger('version_bump');
 * $transaction = $logger->startTransaction('bump_version');
 * $transaction->logEvent('version_change', ['old' => '1.0.0', 'new' => '1.1.0']);
 * $transaction->logSecurityEvent('file_modified', ['file' => 'README.md']);
 * $transaction->end();
 * ```
 */
class AuditLogger
{
    /** @var string Service name */
    private string $service;

    /** @var string User performing actions */
    private string $user;

    /** @var string Directory for audit logs */
    private string $logDir;

    /** @var bool Enable console output */
    private bool $enableConsole;

    /** @var bool Enable file logging */
    private bool $enableFile;

    /** @var int Maximum log file size in MB */
    private int $maxLogSizeMb;

    /** @var int Days to retain audit logs */
    private int $retentionDays;

    /** @var string Session ID */
    private string $sessionId;

    /** @var array Transaction stack */
    private array $transactionStack = [];

    /** @var string Version constant */
    public const VERSION = '03.02.00';

    /**
     * Initialize audit logger.
     *
     * @param string $service Service name (e.g., 'version_bump', 'branch_cleanup')
     * @param string|null $logDir Directory for audit logs (default: logs/audit/)
     * @param string|null $user Username for audit trail (default: from environment)
     * @param bool $enableConsole Output to console (default: true)
     * @param bool $enableFile Write to file (default: true)
     * @param int $maxLogSizeMb Maximum log file size before rotation
     * @param int $retentionDays Days to retain audit logs
     */
    public function __construct(
        string $service,
        ?string $logDir = null,
        ?string $user = null,
        bool $enableConsole = true,
        bool $enableFile = true,
        int $maxLogSizeMb = 10,
        int $retentionDays = 90
    ) {
        $this->service = $service;
        $this->enableConsole = $enableConsole;
        $this->enableFile = $enableFile;
        $this->maxLogSizeMb = $maxLogSizeMb;
        $this->retentionDays = $retentionDays;

        // Determine user
        $this->user = $user ?? $_SERVER['USER'] ?? $_SERVER['USERNAME'] ?? posix_getpwuid(posix_geteuid())['name'] ?? 'unknown';

        // Set up log directory
        if ($logDir === null) {
            // Default to logs/audit/ in repository root
            $repoRoot = dirname(__DIR__, 3);
            $this->logDir = $repoRoot . '/logs/audit';
        } else {
            $this->logDir = $logDir;
        }

        // Create log directory if it doesn't exist
        if ($this->enableFile && !is_dir($this->logDir)) {
            if (!mkdir($this->logDir, 0755, true) && !is_dir($this->logDir)) {
                throw new RuntimeException("Failed to create log directory: {$this->logDir}");
            }
        }

        // Session ID for this logger instance
        $this->sessionId = $this->generateSessionId();

        // Log session start
        $this->logSystemEvent('session_start', [
            'service' => $this->service,
            'user' => $this->user,
            'session_id' => $this->sessionId,
        ]);
    }

    /**
     * Generate unique session ID.
     *
     * @return string Session ID
     */
    private function generateSessionId(): string
    {
        $timestamp = (new DateTime('now', new DateTimeZone('UTC')))->format('Ymd_His');
        $uniqueId = substr(bin2hex(random_bytes(4)), 0, 8);
        return "{$timestamp}_{$uniqueId}";
    }

    /**
     * Generate unique transaction ID.
     *
     * @return string Transaction ID (UUID v4)
     */
    private function generateTransactionId(): string
    {
        return sprintf(
            '%04x%04x-%04x-%04x-%04x-%04x%04x%04x',
            mt_rand(0, 0xffff),
            mt_rand(0, 0xffff),
            mt_rand(0, 0xffff),
            mt_rand(0, 0x0fff) | 0x4000,
            mt_rand(0, 0x3fff) | 0x8000,
            mt_rand(0, 0xffff),
            mt_rand(0, 0xffff),
            mt_rand(0, 0xffff)
        );
    }

    /**
     * Get current log file path with rotation support.
     *
     * @return string Log file path
     */
    private function getLogFilePath(): string
    {
        $dateStr = (new DateTime('now', new DateTimeZone('UTC')))->format('Ymd');
        return "{$this->logDir}/audit_{$this->service}_{$dateStr}.jsonl";
    }

    /**
     * Check if log file should be rotated based on size.
     *
     * @param string $logFile Log file path
     * @return bool True if should rotate
     */
    private function shouldRotateLog(string $logFile): bool
    {
        if (!file_exists($logFile)) {
            return false;
        }

        $sizeMb = filesize($logFile) / (1024 * 1024);
        return $sizeMb >= $this->maxLogSizeMb;
    }

    /**
     * Rotate log file if it exceeds size limit.
     *
     * @param string $logFile Log file path
     */
    private function rotateLogIfNeeded(string $logFile): void
    {
        if ($this->shouldRotateLog($logFile)) {
            $timestamp = (new DateTime('now', new DateTimeZone('UTC')))->format('His');
            $rotatedFile = preg_replace('/\.jsonl$/', ".{$timestamp}.jsonl", $logFile);
            rename($logFile, $rotatedFile);
        }
    }

    /**
     * Write log entry to file and/or console.
     *
     * @param array $entry Log entry data
     */
    private function writeLogEntry(array $entry): void
    {
        // Add timestamp and session info
        $entry['timestamp'] = (new DateTime('now', new DateTimeZone('UTC')))->format('c');
        $entry['session_id'] = $this->sessionId;
        $entry['service'] = $this->service;
        $entry['user'] = $this->user;

        // Console output
        if ($this->enableConsole) {
            $jsonOutput = json_encode($entry, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
            echo "[AUDIT] {$jsonOutput}\n";
        }

        // File output
        if ($this->enableFile) {
            $logFile = $this->getLogFilePath();
            $this->rotateLogIfNeeded($logFile);

            $jsonLine = json_encode($entry, JSON_UNESCAPED_SLASHES) . "\n";
            file_put_contents($logFile, $jsonLine, FILE_APPEND | LOCK_EX);
        }
    }

    /**
     * Log a system event.
     *
     * @param string $eventType Type of system event
     * @param array $data Event data
     */
    private function logSystemEvent(string $eventType, array $data = []): void
    {
        $entry = [
            'event_type' => 'system',
            'event_subtype' => $eventType,
            'data' => $data,
        ];
        $this->writeLogEntry($entry);
    }

    /**
     * Start a new transaction.
     *
     * @param string $operation Operation name
     * @param array $context Additional context data
     * @return AuditTransaction Transaction object
     */
    public function startTransaction(string $operation, array $context = []): AuditTransaction
    {
        $transactionId = $this->generateTransactionId();
        $transaction = new AuditTransaction($this, $transactionId, $operation, $context);
        $this->transactionStack[] = $transactionId;
        return $transaction;
    }

    /**
     * End a transaction.
     *
     * @param string $transactionId Transaction ID to end
     */
    public function endTransaction(string $transactionId): void
    {
        $key = array_search($transactionId, $this->transactionStack, true);
        if ($key !== false) {
            unset($this->transactionStack[$key]);
        }
    }

    /**
     * Log an event within a transaction.
     *
     * @param string $transactionId Transaction ID
     * @param string $eventType Event type
     * @param array $data Event data
     */
    public function logEvent(string $transactionId, string $eventType, array $data = []): void
    {
        $entry = [
            'event_type' => 'audit',
            'transaction_id' => $transactionId,
            'event_subtype' => $eventType,
            'data' => $data,
        ];
        $this->writeLogEntry($entry);
    }

    /**
     * Log a security event.
     *
     * @param string $transactionId Transaction ID
     * @param string $eventType Security event type
     * @param array $data Event data
     */
    public function logSecurityEvent(string $transactionId, string $eventType, array $data = []): void
    {
        $entry = [
            'event_type' => 'security',
            'transaction_id' => $transactionId,
            'event_subtype' => $eventType,
            'severity' => $data['severity'] ?? 'medium',
            'data' => $data,
        ];
        $this->writeLogEntry($entry);
    }
}

/**
 * Audit transaction context manager.
 */
class AuditTransaction
{
    private AuditLogger $logger;
    private string $transactionId;
    private string $operation;
    private array $context;
    private float $startTime;

    public function __construct(
        AuditLogger $logger,
        string $transactionId,
        string $operation,
        array $context = []
    ) {
        $this->logger = $logger;
        $this->transactionId = $transactionId;
        $this->operation = $operation;
        $this->context = $context;
        $this->startTime = microtime(true);

        // Log transaction start
        $this->logger->logEvent($this->transactionId, 'transaction_start', [
            'operation' => $this->operation,
            'context' => $this->context,
        ]);
    }

    /**
     * Get the transaction ID.
     *
     * @return string Transaction ID
     */
    public function getTransactionId(): string
    {
        return $this->transactionId;
    }

    /**
     * Log an event within this transaction.
     *
     * @param string $eventType Event type
     * @param array $data Event data
     */
    public function logEvent(string $eventType, array $data = []): void
    {
        $this->logger->logEvent($this->transactionId, $eventType, $data);
    }

    /**
     * Log a security event within this transaction.
     *
     * @param string $eventType Security event type
     * @param array $data Event data
     */
    public function logSecurityEvent(string $eventType, array $data = []): void
    {
        $this->logger->logSecurityEvent($this->transactionId, $eventType, $data);
    }

    /**
     * End the transaction.
     *
     * @param string|null $status Transaction status (success|failure)
     * @param array $result Transaction result data
     */
    public function end(?string $status = 'success', array $result = []): void
    {
        $duration = microtime(true) - $this->startTime;

        $this->logger->logEvent($this->transactionId, 'transaction_end', [
            'operation' => $this->operation,
            'status' => $status,
            'duration_seconds' => round($duration, 3),
            'result' => $result,
        ]);

        $this->logger->endTransaction($this->transactionId);
    }
}
